saltgui/static/scripts/routes/grains.js

```diff
@@ -0,0 +1,242 @@
+class GrainsRoute extends PageRoute {
+
+  constructor(router) {
+    super("^[\/]grains$", "Grains", "#page_grains", "#button_grains");
+    this.router = router;
+    this.keysLoaded = false;
+    this.jobsLoaded = false;
+
+    this._showGrains = this._showGrains.bind(this);
+    this._updateJobs = this._updateJobs.bind(this);
+    this._updateKeys = this._updateKeys.bind(this);
+    this._updateMinion = this._updateMinion.bind(this);
+
+    // These grains are predefined
+    this.knownKeys = [
+      // Linux
+      "biosreleasedate",
+      "biosversion",
+      "cpuarch",
+      "cpu_flags",
+      "cpu_model",
+      "disks",
+      "dns",
+      "domain",
+      "fqdn",
+      "fqdn_ip4",
+      "fqdn_ip6",
+      "gid",
+      "gpus",
+      "groupname",
+      "host",
+      "hostname",
+      "hwaddr_interfaces",
+      "id",
+      "init",
+      "ip4_gw",
+      "ip4_interfaces",
+      "ip6_gw",
+      "ip6_interfaces",
+      "ip_gw",
+      "ip_interfaces",
+      "ipv4",
+      "ipv6",
+      "kernel",
+      "kernelrelease",
+      "kernelversion",
+      "locale_info",
+      "localhost",
+      "lsb_distrib_codename",
+      "lsb_distrib_id",
+      "machine_id",
+      "manufacturer",
+      "master",
+      "mdadm",
+      "mem_total",
+      "nodename",
+      "num_cpus",
+      "num_gpus",
+      "os",
+      "osarch",
+      "oscodename",
+      "os_family",
+      "osfinger",
+      "osfullname",
+      "osmajorrelease",
+      "osrelease",
+      "osrelease_info",
+      "path",
+      "pid",
+      "productname",
+      "ps",
+      "pythonexecutable",
+      "pythonpath",
+      "pythonversion",
+      "saltpath",
+      "saltversion",
+      "saltversioninfo",
+      "selinux",
+      "serialnumber",
+      "server_id",
+      "shell",
+      "SSDs",
+      "swap_total",
+      "systemd",
+      "uid",
+      "username",
+      "uuid",
+      "virtual",
+      "zfs_feature_flags",
+      "zfs_support",
+      "zmqversion",
+      // Additional keys for Windows
+      "motherboard",
+      "osmanufacturer",
+      "osservicepack",
+      "osversion",
+      "timezone",
+      "windowsdomain",
+      // Additional keys from Salt Source Code
+      "computenode_sdc_version",
+      "computenode_vm_capable",
+      "computenode_vm_hw_virt",
+      "computenode_vms_running",
+      "computenode_vms_stopped",
+      "computenode_vms_total",
+      "computenode_vms_type",
+      "datacenter",
+      "fc_wwn",
+      "hypervisor_uuid",
+      "imageversion",
+      "iscsi_iqn",
+      "lsb_distrib_release",
+      "manufacture",
+      "mdata",
+      "osbuild",
+      "osrelease_stamp",
+      "pkgsrcpath",
+      "pkgsrcversion",
+      "product",
+      "roles",
+      "systemfirmware",
+      "systemfirmwaredate",
+      "virtual_subtype",
+      "zpool",
+    ];
+  }
+
```

Comment:
Isn't there a api call to retrieve a list of available grains?

Reply:
Searched and found nothing about that in the API. Then did a source code check to find them all, including those that are only returned under certain circumstances. A bridge too far?

Reply:
Don't think it's a bridge too far as it it's a really handy feature. I dislike the fact that we need to hardcode a grain list while this is depending on the salt version, and that grains can be custom as well.

```grains.ls``` will show the grains right? isn't there something we can do?

Reply:
`grains.ls` shows all grains whether predefined or custom

Reply:
`grains.ls shows all grains whether predefined or custom.`

That's a plus right? :) What would be the problem with  that?

Reply:
I missed the fact that there was confusion at the start of this thread.
The list exists *only* to distinguish the built-in grains from the custom grains. If that distinction is not worth the trouble, then it can go. My remark *a bridge too far* was only about using the list to distinguish the built-in grains from the custom grains.